src/Serialization/MinFee.purs

```diff
@@ -44,16 +48,34 @@ calculateMinFeeCsl (ProtocolParameters pparams) txNoSigs = do
       BigNum.toBigInt (_minScriptFee exUnitPricesCsl cslTx)
   pure $ wrap $ minFee + minScriptFee + BigInt.fromInt 3 * txFeePerByte
 
+-- | Adds fake signatures for each expected signature of a transaction.
 addFakeSignatures :: T.Transaction -> T.Transaction
 addFakeSignatures tx =
   let
-    nRequiredSigners =
+    -- requiredSigners field of a transaction
+    requiredSigners :: Set Ed25519KeyHash
+    requiredSigners =
       tx # unwrap >>> _.body >>> unwrap >>> _.requiredSigners
-        >>> map Array.length
-        >>> fromMaybe 1
+        >>> fromMaybe mempty
+        >>> map unwrap
+        >>> Set.fromFoldable
+
+    -- All possible signers from NativeScript.
+    nsPossibleSigners :: Int
+    nsPossibleSigners = getMaximumSigners requiredSigners $ ScriptAll
+      ( tx # unwrap >>> _.witnessSet >>> unwrap >>> _.nativeScripts >>>
+          fromMaybe mempty
+      )
+
+    -- We want to add space for required signatures (at least one, if
+    -- none specified).
+    nRequiredSigners = tx # unwrap >>> _.body >>> unwrap >>> _.requiredSigners
+      >>> map (map unwrap >>> Array.length)
+      >>> fromMaybe 1
+
   in
     tx # _witnessSet <<< _vkeys .~ Just
-      (Array.replicate nRequiredSigners fakeVkeywitness)
+      (Array.replicate (nRequiredSigners + nsPossibleSigners) fakeVkeywitness)
```

Comment:
If `requiredSigners` is empty, but `nsPossibleSigners` isn't, will this count one too many, owing to the `fromMaybe 1` above? Is this even possible?

Reply:
It depends on what's inside the script. We need to add at least one witness, but if we don't know the expected signer, we can't know if it is used in the script. So it's the best we can do.